src/PrestaShopBundle/Translation/PrestaShopTranslatorTrait.php

```diff
@@ -53,27 +53,38 @@ public function trans($id, array $parameters = [], $domain = null, $locale = nul
             $legacy = $parameters['legacy'];
             unset($parameters['legacy']);
         }
+        $emptyParams = empty($parameters);
+        $isSprintf = $this->isSprintfString($id);
 
         if (empty($locale)) {
             $locale = null;
         }
 
-        $translated = parent::trans($id, [], $this->normalizeDomain($domain), $locale);
+        if (!$emptyParams && $isSprintf) {
+            $translated = parent::trans(
+                $id,
+                 [],
+                $this->normalizeDomain($domain),
+                $locale
+            );
+        } else {
+            $translated = parent::trans(
```

Comment:
This breaks legacy fallback which should also work without parameters.

I understand the subject is quite tricky. Let me try to clarify.

The translator contais all the base translations + all the ones customized by the user using the BO interface, but it doesn't contain the translations being loaded from legacy dictionaries (which cannot be loaded inside the translator because they are incompatibile).  

When the translation exists in the translator, the message is translated and that's the end of it. But when the translation doesn't exist in the translator, then we try and perform it using the legacy module dictionaries. As you can see in `shouldFallbackToLegacyModuleTranslation()`, we use `$message === $translated` to test if the translation using the `trans()` was unsuccessful. 

However, we can't use that test if you send parameters to `trans()` because the translator will replace the parameters in the output even if it fails to find a translation. Eg: `trans("message %test%", ['%test%' => "toto"])` will yield `"message toto"`, which is different from the original "message %test%".

You'll probably say that comparing the original message to the output to infer behavior is hacky, and I would agree. The only way to know for sure would be to look inside its catalogues for a match.

I'm pretty sure `has()` will return false when the translation is not found (but you should check to make sure). This is pretty obvious, so I'm suspicious of why we didn't do this before, perhaps I'm missing something.

If all goes well, you should end up with a code that looks like this

```php
$normalizedDomain = $this->normalizeDomain($domain);
$translationExists = $this->getCatalogue($locale)->has($id, $normalizedDomain);

if ($this->shouldFallbackToLegacyModuleTranslation($id, $domain)) {
    return $this->translateUsingLegacySystem($id, $parameters, $domain, $locale);
}

if (!empty($parameters) && !$this->isSprintfString($id)) {
    // translate with parameters
    $translated = parent::trans(
        $id,
        $parameters,
        $normalizedDomain,
        $locale
    );
} else {
    // translate without parameters, 
    $translated = parent::trans(
        $id,
        [],
        $normalizedDomain,
        $locale
    );

    if (!empty($parameters)) {
        $translated = vsprintf($translated, $parameters);
    }
}

if (isset($legacy) && is_callable($legacy)) {
    if ('htmlspecialchars' === $legacy) {
        $translated = call_user_func($legacy, $translated, ENT_NOQUOTES);
    } else {
        $translated = call_user_func($legacy, $translated);
    }
}

return $translated;

// [ ... ]

private function shouldFallbackToLegacyModuleTranslation($message, $domain)
{
    return 
        'Modules.' === substr($domain ?? '', 0, 8);
        && (
            !method_exists($this, 'getCatalogue')
            || !$this->getCatalogue()->has($message, $this->normalizeDomain($domain))
        )
        ;
}
```

Reply:
Hi @eternoendless your comment was probably made on the previous commit, the latest state has greatly simplified the previous code and it meets nearly all what you suggested in a smaller way except for a few things.

`shouldFallbackToLegacyModuleTranslation` is based on the translated message it checks that the initial message hasn't changed, what you suggest is to use legacy method early but I'm not sure about this one Because if the message is absent in module's legacy translations then the parameters could never be replaced, although your idea to accurately detect if the legacy translation exists makes sense, but we'd need a `LegacyTranslator::has` method as well, from what I could grasp the existence is checked in this code:

https://github.com/PrestaShop/PrestaShop/blob/d0f9edeb15b7c1b249d8b7fdb692987d3270f6f2/classes/Translate.php#L187-L197

If we can extract this in a `has` method then we can really check if the message is handled via legacy (also we should check that it's NOT handled by the catalogue since it should have the priority) and when we have this then I think it's the most optimal fallback we can have.

Also in your suggestion the `$translationExists` is never used, but it should be handled in the `shouldFallbackToLegacyModuleTranslation` method I guess (seems like it already is)

Reply:
```php
if ($this->shouldFallbackToLegacyModuleTranslation($id, $domain)) {
    return $this->translateUsingLegacySystem($id, $parameters, $domain, $locale);
}

if (!empty($parameters) && !$this->isSprintfString($id)) {
    // translate with parameters
    $translated = parent::trans(
        $id,
        $parameters,
        $this->normalizeDomain($domain),
        $locale
    );
} else {
    // translate without parameters, 
    $translated = parent::trans(
        $id,
        [],
        $this->normalizeDomain($domain),
        $locale
    );

    if (!empty($parameters)) {
        $translated = vsprintf($translated, $parameters);
    }
}

if (isset($legacy) && is_callable($legacy)) {
    if ('htmlspecialchars' === $legacy) {
        $translated = call_user_func($legacy, $translated, ENT_NOQUOTES);
    } else {
        $translated = call_user_func($legacy, $translated);
    }
}

return $translated;

// [ ... ]

private function shouldFallbackToLegacyModuleTranslation($message, $domain)
{
    return 
        'Modules.' === substr($domain ?? '', 0, 8);
        && (
            !method_exists($this, 'getCatalogue')
            || !$this->getCatalogue()->has($message, $this->normalizeDomain($domain))
        )
        ;
}
```

The legacy translator has its own way of replacing params, so no need to add more logic: https://github.com/PrestaShop/PrestaShop/blob/434137b764ab0faa9de420a0a8ec54a039165dd0/classes/Translate.php#L199-L205